### PR TITLE
Add tag regex

### PR DESCRIPTION
**Pull request by angersock**
## Description

This is a quick attempt at putting in a regex-based tag suggestion system for Lobsters.

## Motivation

If a user submits a story whose title is "Announcing Linux 5.3!", we should be able to guess that the tags relevant may be `release` and `linux`. Similarly, a submission of "Foobar for Python and Go" is pretty obviously likely to be about `python` and `go`. 

We can nudge users in the correct direction by providing feedback *before* they make mislabeled submissions. I think this can be done with a simple regex-based approach without having to do ML or stats or anything.

## Implementation

It works by adding a "suggestion_regex" field to the tag model, and then changing the story submission form to have a `regex_suggested_tags` div that is normally empty but, after a run of the `checkStoryTitle()` function, will be populated with a list of tags whose regexes matched on the title.

Any regex of zero length is considered as "do not try to suggest this tag automagically".

## Future improvements

* Adding more robust protection against malicious tag regexes. No limit in size is currently present, and no attempt is made on the server-side that the submitted regex is valid for JS. No attempt is made to prevent input of, say, a regex that is just whitespace.
* Adding a way of testing the regex in the tag edtior.
* Changing the tag suggestions to be clickable (this would almost certainly be a big help).
* Preventing display of tags that have already been selected.
* Tests. Always tests.

~

I'm throwing this up for @hmadison to kick the tires on, and we can proceed from there.

### CODE REVIEW

File: db/migrate/20190916000000_add_tag_regex.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class AddTagRegex < ActiveRecord::Migration[5.2]
+    def change
+      add_column :tags, :suggestion_regex, :string, :default => ""
```
[the remark this answers is not in the document]
Comment (angersock) [reply db/migrate/20190916000000_add_tag_regex.rb]: That's exactly it--just to save on handling extra cleanups for existing records that wouldn't already have it. I've been bitten by non-null constraints before, so I think that'd be a good "future improvements" addition. :+1:

File: app/models/tag.rb [hunk 2]
```diff
@@ -10,7 +10,7 @@ class Tag < ApplicationRecord
 
   after_save :log_modifications
 
-  attr_accessor :edit_user_id, :stories_count
+  attr_accessor :edit_user_id, :stories_count, :suggestion_regex
```
[the remark this answers is not in the document]
Comment (angersock) [reply app/models/tag.rb]: Fixed, ty ty.

**Comment by swifthand** [comment]
Two thoughts on the first bullet point under "future improvements":

>Adding more robust protection against malicious tag regexes. No limit in size is currently present, and no attempt is made on the server-side that the submitted regex is valid for JS. No attempt is made to prevent input of, say, a regex that is just whitespace.

### Length Limits

The use of the `string` type rather than `text` in ActiveRecord migrations will end up with an implicit limit of 255 in most cases (i.e. becomes a `character varying(255)` in PostgreSQL). So you do have something there, it's just going to cause a 500 error (unless cause elsewhere, of course) when the DB throws up with (in PostgreSQL):
```
ActiveRecord::StatementInvalid: PG::StringDataRightTruncation: ERROR:  value too long for type character varying(255)
```

Rather than this, a more user-friendly validation error could be achieved via: 
```ruby
validates :suggestion_regex, length: { maximum: 255 }
```

### Validate JavaScript Regex from Ruby

For your consideration:

```ruby
validate :suggestion_regex_valid_js?

def suggestion_regex_valid_js?
  unless suggestion_regex.blank? || valid_js_regex?(suggestion_regex)
    errors.add(:suggestion_regex, "is not a valid JavaScript RegExp")
  end
end

def valid_js_regex?(str)
  encoded_str = Base64.strict_encode64(str)
  regex_validation = %{
    (function isValidRegex(input) { 
      try {
        new RegExp(input);
        return true;
      } catch(e) {
        return false;
      } 
    })(new Buffer('#{encoded_str}', 'base64').toString())
  }
  ExecJS.eval(test_function)
rescue ExecJS::RuntimeError, ExecJS::ProgramError
  false
end
```

And some notes on this:

1. I prefer adding validators via implementing `ActiveModel::EachValidator` classes, but I do not see any precedent for that in the existing Lobsters code. So instead this is written to use a one-off `validate :some_instance_method` in the model.
2. I would be disappointed if passing user-provided strings into nodejs on the server didn't raise someone's hackles! So strict Base64 encoding is used to attempt to avoid exploitation by passing the regex along in a format that is as benign as possible. Security-centric criticism is welcomed.
3. Performing the Base64 decoding via `Buffer` adds an implicit dependency a Nodejs version, but to the best of my knowledge any version from at least 2015 should be fine. Or even earlier? I am one-hundred percent sure that Lobsters should be fine giving that it is running Rails 5.
4. The entire `valid_js_regex?` method could easily be done in one single expression. However, for unconventional code such as this I wanted to make each little bit utterly explicit. Hence the intermediate variable assignments.
5. I skimmed through `ExecJS` and did not see any other exceptions that it might raise, other than the two listed here, or possibly an encoding error, which ought not (?) to reach the model in the first place (i.e. assuming utf-8 is enforced elsewhere). Someone sanity check me on that. 
6. Although I am somewhat confident in this implementation, I would unit test the hell out of this just to be sure :sweat_smile:

**Comment by numinit** [comment]
https://github.com/discourse/mini_racer#fork-safety

Will this matter for deploying lobsters?

**Comment by numinit** [comment]
Actually, let's sidestep this issue entirely and use duktape. The V8 runtime uses a ton of extra resources by comparison.

# Experiment

Add one of the following to the Gemfile:

```
gem "mini_racer", "~> 0.2"
gem "duktape", "~> 2.3.0.0"
```

Then, follow these steps:

1. Run `bundle ex rails c` to drop into a console.
2. puts `cat /proc/#{Process.pid}/status | egrep 'VmSize|VmRSS|Threads'`
3. ` ExecJS.compile('function isValidRegexp(x) {try {return !!new RegExp(x);} catch (e) {return false;}}').('isValidRegexp', '^http:\/\/|^https:\/\/')`
4. Repeat step 2

## duktape

```
### Pre-ExecJS
VmSize: 672636 kB
VmRSS:  82496 kB
Threads:        8
```

### Post-ExecJS

```
VmSize: 674520 kB
VmRSS:  84600 kB
Threads:        8
```

## mini_racer

### Pre-ExecJS

```
VmSize: 689492 kB
VmRSS:  86664 kB
Threads:        8
```

### Post-ExecJS

```
VmSize: 867516 kB
VmRSS:  94076 kB
Threads:        13
```

In short, Duktape doesn't spawn any extra threads, and has less memory allocation footprint. It [supports](https://github.com/svaarala/duktape/blob/master/doc/regexp.rst) "standard" ECMAScript regexes, too.

**Comment by pushcx** [comment]
I'm going to close this out. Adding node to make this testable is a huge amount of complexity for a small feature. Even without that, this is sort of a feature that makes an infinite amount of new work tweaking regexps for each tag and each new story, and the pain it addresses of mis-tagging is pretty minor.
